[www] Change paywall logic to continuously poll.

sndurkin opened the pull request:
This PR changes the paywall logic to continuously watch the paywall addresses for unpaid users. The full behavior is:

* When a user verifies his registration, his paywall info is automatically added to the verification pool.
* The verification pool iterates the unpaid users, checking dcrdata and/or insight for a transaction that meets the paywall requirements.
* When a user is verified as paid by the pool, it removes his paywall info from the pool and updates the database.
* When the server first starts, it iterates all users in the database and adds unpaid users to the pool.
* The pool will stop watching any users whose paywall info was generated more than 24 hours ago. The user's paywall info will only be re-added to the pool via a call to `/me` or `/login` or `/user/verifypayment` (and the timer resets).
* Calls to `/me`, `/login`, and `/user/verifypayment` will check the database for the user's paywall status.

**Note:** The `/user/verifypaymenttx` route has changed to `/user/verifypayment` and no longer accepts a transaction id, because the server performs its own search for a suitable transaction.

Closes #313.

chappjc commented on politeiawww/paywall.go:
```diff
@@ -1,47 +1,256 @@
+			if b.HasUserPaid(user) {
+				// The user could have been marked as paid by RouteVerifyUserPayment,
+				// so just remove him from the in-memory pool.
+				mutex.Lock()
```
Looks like you are trying to upgrade a read lock to a read/write lock.  Go doesn't do that.  You'll have to RUnlock first

chappjc commented on politeiawww/paywall.go:
```diff
@@ -1,47 +1,256 @@
+			}
+
+			if paywallHasExpired(user.NewUserPaywallTxNotBefore) {
+				return
```
Still locked

sndurkin commented:
Thanks @chappjc, I've addressed the issues you pointed out; I was indeed thinking I could upgrade to a `RWLock`, and the `return` statement should have been `continue`.

marcopeereboom commented on politeiawww/paywall.go:
```diff
@@ -1,47 +1,280 @@
+)
+
+var (
+	paywallUsers map[uint64]paywallInfo
```
All this stuff needs to go in the context and not be global. Also use the mutex that already exists in the context instead of making a new one.

marcopeereboom commented on politeiawww/paywall.go:
```diff
@@ -1,47 +1,280 @@
+func (b *backend) checkForPayments() {
+	minConfirmations := b.cfg.MinConfirmationsRequired
+
+	mutex.Lock()
```
That is a lot of locking gymnastics. If these are relatively fast operations just take the lock and defer the unlock. If they are slow then we need to do some gymnastics. Like break the function apart and do the bits that need locking in a single function; reason for that is that lcoking doesn't become intertwined like it is now. If someone adds a return we'll have a deadlock.

Can we do the init call in a more generic init area? pretty sure we shouldn't be calling the init all the time.

marcopeereboom commented on politeiawww/paywall.go:
```diff
@@ -1,47 +1,280 @@
 }
+
+// AddUserToPaywallPool adds a user and its paywall info to the in-memory pool.
+func (b *backend) AddUserToPaywallPool(user *database.User) {
```
Add context mutex and mark the doco with "Must be called WITHOUT the mutex held."

marcopeereboom commented on politeiawww/paywall.go:
```diff
@@ -1,47 +1,280 @@
+	}
+
+	// Create the in-memory pool of all users who need to pay the paywall.
+	mutex.Lock()
```
Deadlock if there is any failure.

Per last comment, make the bits that need to be locked into a function if possible. Much easier to keep track of locking.

Add doco and mark that the call must be called WITHOUT the mutex held.

marcopeereboom commented on politeiawww/paywall.go:
```diff
@@ -1,47 +1,280 @@
+
+	// paywallCheckGap is the amount of time the server sleeps after polling for
+	// a paywall address.
+	paywallCheckGap = time.Second * 1
```
That seems overly aggressive.

marcopeereboom commented on politeiawww/paywall.go:
```diff
@@ -1,47 +1,318 @@
+	b.RLock()
+	defer b.RUnlock()
+
+	copy := make(map[uint64]paywallPoolMember, len(b.userPaywallPool))
```
copy is a builtin. Please rename. And len on map???
